src/common/modules/AutocorrectHandler.js

```diff
@@ -0,0 +1,211 @@
+                    index = aindex;
+                    length = y.length;
+                } else if (aindex === index && y.length > length) {
+                    length = y.length;
+                }
+            }
+        }
+
+        if (length > 0) {
+            length = x.length - (index + length);
+            if (length > 1) {
+                apatterns.push(x.slice(0, -(length - 1)));
+            }
+        }
+    }
+    apatterns = apatterns.filter((item, pos) => apatterns.indexOf(item) === pos);
+    console.log("Do not autocorrect for these patterns", apatterns);
+
+    apatterns.forEach((symbol, index) => {
+        apatterns[index] = symbol.replace(re, "\\$&");
+    });
+
+    symbolpatterns = new RegExp(`(${symbolpatterns.join("|")})$`);
+    apatterns = new RegExp(`(${apatterns.join("|")})$`);
+}
+
+/**
+ * On error.
+ *
+ * @param {string} error
+ * @returns {void}
+ */
+function onError(error) {
+    console.error(`Error: ${error}`);
+}
+
+/**
+ * Set autocorrect settings.
+ *
+ * @param {Object} autocorrect
+ * @returns {void}
+ */
+function setSettings(autocorrect) {
+    settings.autocorrectEmojis = autocorrect.autocorrectEmojis;
+    settings.autocorrectEmojiShortcodes = autocorrect.autocorrectEmojiShortcodes;
+    settings.autocorrectSymbols = autocorrect.autocorrectSymbols;
+    settings.autocomplete = autocorrect.autocompleteEmojiShortcodes;
+    settings.quotes = autocorrect.autocorrectUnicodeQuotes;
+    settings.fracts = autocorrect.autocorrectUnicodeFracts;
+
+    applySettings();
+}
+
+/**
+ * Send autocorrect settings to content scripts.
+ *
+ * @param {Object} autocorrect
+ * @returns {void}
+ */
+function sendSettings(autocorrect) {
+    setSettings(autocorrect);
+
+    browser.tabs.query({}).then((tabs) => {
+        for (const tab of tabs) {
+            browser.tabs.sendMessage(
+                tab.id,
+                {
+                    "type": COMMUNICATION_MESSAGE_TYPE.AUTOCORRECT_CONTENT,
+                    "autocomplete": settings.autocomplete,
+                    "quotes": settings.quotes,
+                    "fracts": settings.fracts,
+                    "autocorrections": autocorrections,
+                    "longest": longest,
+                    "symbolpatterns": symbolpatterns,
+                    "apatterns": apatterns,
+                    "emojiShortcodes": emojiShortcodes
+                }
+            ).catch(onError);
+        }
+    }).catch(onError);
+}
+
+/**
+ * Init autocorrect module.
+ *
+ * @public
+ * @returns {void}
+ */
+export async function init() {
+    const autocorrect = await AddonSettings.get("autocorrect");
+
+    for (const key in emojiMart.emojiIndex.emojis) {
+        const emoji = emojiMart.emojiIndex.emojis[key];
+        emojiShortcodes[emoji.colons] = emoji.native;
+    }
+
+    Object.freeze(emojiShortcodes);
+
+    setSettings(autocorrect);
+
+    browser.runtime.onMessage.addListener((message, sender, sendResponse) => {
```

Comment:
Hmm, why do custom coding here. Can't we just use `BrowserCommunication.addListener`?

Reply:
No, because it needs to be able to send a response and `BrowserCommunication` does not support that.

Reply:
Actually it should be possible: https://github.com/TinyWebEx/BrowserCommunication/blob/19e9abc44bac6724acebfe355da8aeb1e3b424c8/BrowserCommunication.js#L25-L38

`sendResponseCallback` is exposed there.

Reply:
Oops, yes, you are correct. It is needed because the message is coming from the content script, which cannot use modules like `BrowserCommunication`.

Reply:
Hmm, anyway, this only means the content script needs to manually specify `COMMUNICATION_MESSAGE_TYPE.AUTOCORRECT_CONTENT`, everything else should be compatible here. :smiley:

Reply:
Hmm, I tried updating it to use `BrowserCommunication`, but it did not work. The response is never received by the content script.